### PR TITLE
Add counterparty support

### LINKED ISSUE #34: Transactions List (with expand_merchants=True) fails pydantic checks
Running `monzo_api.transactions.list(account_id=account_id, expand_merchant=True)` produces pydantic errors:

```
pydantic_core._pydantic_core.ValidationError: 2 validation errors for MonzoTransaction
merchant.MonzoTransactionMerchant.suggested_tags
  Field required [type=missing, input_value={'id': 'merch_0000ADdaX06...: False, 'metadata': {}}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.7/v/missing
merchant.str
  Input should be a valid string [type=string_type, input_value={'id': 'merch_0000ADdaX06...: False, 'metadata': {}}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.7/v/string_type
```

These look like they could be tied together...  I don't see a 'suggested_tags' option in any of the transactions I have.

### PR DESCRIPTION
**Pull request by csogilvie**
Also fixes #34

### CODE REVIEW

**Comment by csogilvie** [comment]
I note that the tests are failing in some cases, but my local version was doing this before the code changes I made too.

**Comment by pawelad** [comment]
Thank you for taking the time to do this! Next time, would you mind separating the bug fix from the new feature (as in, one PR per bug / feature / thing)?

Don't worry about the failing tests CI - unfortunately I wasn't able to get it to work on PRs and decided to leave it for now, as I think I want to get rid of VCR usage anyway, it's too fragile with how undocumented the Monzo API is.

**Comment by csogilvie** [comment]
> Thank you for taking the time to do this! Next time, would you mind separating the bug fix from the new feature (as in, one PR per bug / feature / thing)?

Can do; sorry - was just the way I've been itteratively fixing things locally to get things to work!

**Comment by pawelad** [comment]
Also, do you mind sharing the raw JSON for an example transaction with the `counterparty` set? I don't see it in their docs nor in any of my transactions. Thanks!

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,11 @@ The format is based on [Keep a Changelog], and this project adheres to
 [Semantic Versioning].
 
 ## Unreleased
+### Fixed
+- Transactions List (with expand_merchants=True) fails pydantic Checks (Issue [#34](https://github.com/pawelad/pymonzo/issues/34))
```
Comment (pawelad) [inline CHANGELOG.md]:
```suggestion
- Listing transactions with `expand_merchants=True`  fails when `suggested_tags` aren't present (by [@csogilvie](https://github.com/csogilvie)) [[#34](https://github.com/pawelad/pymonzo/issues/34)]
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,11 @@ The format is based on [Keep a Changelog], and this project adheres to
...
+
+### Added
+- Add Counterparty Support to Monzo Transactions (by [@csogilvie](https://github.com/csogilvie))
```
Comment (pawelad) [inline CHANGELOG.md]:
```suggestion
- Add `counterparty` field support to Monzo transactions (by [@csogilvie](https://github.com/csogilvie)) [[#34](https://github.com/pawelad/pymonzo/issues/34)]
```

File: src/pymonzo/transactions/schemas.py [hunk 3]
```diff
@@ -94,7 +94,7 @@ class MonzoTransactionMerchant(BaseModel):
     online: bool
     atm: bool
     disable_feedback: bool
-    suggested_tags: str
+    suggested_tags: Optional[str] = None
```
Comment (pawelad) [inline src/pymonzo/transactions/schemas.py]: Do you mind putting this below `metadata` please, so the attributes with default values are at the end? Thanks!

File: src/pymonzo/transactions/schemas.py [hunk 4]
```diff
@@ -122,6 +122,41 @@ def __rich__(self) -> Table:
...
+        account_number: The account number of the other party
+        name: The name of the other party
+        sort_code: The sort code of the other party
+        user_id: Monzo internal User ID of the other party
+    """
+
+    account_number: Optional[str] = None
+    name: Optional[str] = None
+    sort_code: Optional[str] = None
+    user_id: Optional[str] = None
```
Comment (pawelad) [inline src/pymonzo/transactions/schemas.py]: Should these be optional? I'd expect `counterparty` to be either set to `None` or have all these present, no?

File: src/pymonzo/transactions/schemas.py [hunk 5]
```diff
@@ -162,6 +197,7 @@ class MonzoTransaction(BaseModel):
     description: str
     id: str
     merchant: Union[MonzoTransactionMerchant, str, None]
+    counterparty: Union[MonzoTransactionCounterparty, dict, None] = None
```
Comment (pawelad) [inline src/pymonzo/transactions/schemas.py]: Do you mind moving this below `settled` please, alongside other attributes with default values? Thanks!

Also, should we accept `dict` here? Shouldn't be either in the format specified in `MonzoTransactionCounterparty` or `None`?

File: src/pymonzo/transactions/schemas.py [hunk 6]
```diff
@@ -175,6 +211,12 @@ def empty_str_to_none(cls, v: str) -> Optional[str]:
         """Convert empty strings to `None`."""
         return empty_str_to_none(v)
 
+    @field_validator("counterparty", mode="before")
+    @classmethod
+    def empty_dict_to_none(cls, v: dict) -> Optional[dict]:
+        """Convert empty dict to `None`."""
+        return empty_dict_to_none(v)
```
Comment (pawelad) [inline src/pymonzo/transactions/schemas.py:220]: So do some transactions have `counterparty` set to `None`, some to an empty dict `{}` and some to the schema specified in `MonzoTransactionCounterparty`?

**Review by pawelad — CHANGES_REQUESTED** [review]

**Comment by csogilvie** [comment]
> Also, do you mind sharing the raw JSON for an example transaction with the `counterparty` set? I don't see it in their docs nor in any of my transactions. Thanks!

`{
	account_id: acc_id,
	amount: amt,
	amount_is_pending: false,
	atm_fees_detailed: null,
	attachments: null,
	can_add_to_tab: false,
	can_be_excluded_from_breakdown: false,
	can_be_made_subscription: false,
	can_match_transactions_in_categorization: false,
	can_split_the_bill: false,
	categories: {
		income: amt
	},
	category: "income",
	counterparty: {
		account_number: "12345678",
		name: "name",
		sort_code: "123456",
		user_id: "anonuser_string"
	},
	created: "2024-04-21T05:30:14.08Z",
	currency: "GBP",
	dedupe_id: "ID",
	description: "DESC",
	fees: { },
	id: "tx_string",
	include_in_spending: false,
	international: null,
	is_load: false,
	labels: null,
	local_amount: amt,
	local_currency: "GBP",
	merchant: null,
	merchant_feedback_uri: "URL",
	metadata: {
		faster_payment: "true",
		fps_fpid: "fp_id",
		fps_payment_id: "payment_id",
		insertion: "id",
		ledger_committed_timestamp_earliest: "2024-04-21T05:30:14.336Z",
		ledger_committed_timestamp_latest: "2024-04-21T05:30:14.336Z",
		notes: "notes",
		standin_correlation_id: "id",
		trn: "id"
	},
	notes: "notes",
	originator: false,
	parent_account_id: "",
	scheme: "payport_faster_payments",
	settled: "2024-04-22T06:00:00Z",
	updated: "2024-04-21T05:30:14.801Z",
	user_id: ""
}`

File: src/pymonzo/transactions/schemas.py (hunk 4 above)
Comment (csogilvie) [reply src/pymonzo/transactions/schemas.py]: counterparty either comes through as:

`counterparty: {}`

or `counterparty: {
		account_number: "",
		name: "",
		sort_code: "",
		user_id: ""
	}`

and this is what worked for me... although happy to look again :)

File: src/pymonzo/transactions/schemas.py (hunk 6 above)
Comment (csogilvie) [reply src/pymonzo/transactions/schemas.py:220]: Either the empty dict or the schema specified in MonzoTransactionCounterparty... 

I've not used pydantic before so maybe there are better ways of achieving what I've done...

File: src/pymonzo/transactions/schemas.py (hunk 4 above)
Comment (csogilvie) [reply src/pymonzo/transactions/schemas.py]: I think I've cleaned this up now...

File: src/pymonzo/transactions/schemas.py (hunk 5 above)
Comment (csogilvie) [reply src/pymonzo/transactions/schemas.py]: I think I've cleaned this up now...

File: src/pymonzo/transactions/schemas.py (hunk 6 above)
Comment (csogilvie) [reply src/pymonzo/transactions/schemas.py:220]: I think I've cleaned this up now...

File: src/pymonzo/transactions/schemas.py (hunk 4 above)
Comment (pawelad) [reply src/pymonzo/transactions/schemas.py]: Well, that's not great on Monzo's side... OK, let me have a quick look into this and get back to you.

File: src/pymonzo/transactions/schemas.py (hunk 6 above)
Comment (pawelad) [reply src/pymonzo/transactions/schemas.py:220]: Well, that's not great on Monzo's side... OK, let me have a quick look into this and get back to you.

**Review by pawelad — APPROVED** [review]
Thanks for this!

The `empty_dict_to_none` util didn't actually work as expected, because it returned `None` even if the passed value was a valid `MonzoTransactionCounterparty` instance so I took the liberty to fix it.

**Pull request merged**